### PR TITLE
blobs: assorted bug fixes

### PR DESCRIPTION
**Pull request by knz**
See the individual commits for details.

### CODE REVIEW

**Review by ajwerner — APPROVED** [review]
Thanks for doing this. I don't fundamentally disagree with anything you've done in this PR but it seems like a really bad idea to be using TMPDIR here at all. 

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42542#-:-LtziZa3A9CAg8wOYz2C:b8qxc0n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @g3orgia)

___
*[pkg/blobs/local_storage.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42542#-LtzhxVM3BBG27_qj_Ht:-LtzhxVN52Y2zvl-iBsd:bftk3sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4801e3f79f73270d9b51f7ecf25b884955c8e64/pkg/blobs/local_storage.go#L69)):*
> ```Go
> 		return err
> 	}
> 	tmpFile, err := ioutil.TempFile(os.TempDir(), filepath.Base(fullPath))
> ```

My sense is that while handling the cross device errors is a good idea, we probably shouldn't be writing the temp file into os.TempDir() but rather we should be writing it to the external IO dir. This would eliminate cross device IO except in cases where users have subdirectories of their external IO dir on different devices. 

Writing these files to TMPDIR  will potentially cause lots of problems for our users. We expect the external io dir to be big enough to hold files but I don't expect users to anticipate that they need adequate space in their TMPDIR.
___
*[pkg/blobs/local_storage.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42542#-LtziagREIvf6mIEXeVu:-LtziagREIvf6mIEXeVv:b-4x7iw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4801e3f79f73270d9b51f7ecf25b884955c8e64/pkg/blobs/local_storage.go#L73)):*
> ```Go
> 		return errors.Wrap(err, "creating tmp file")
> 	}
> 	// TODO(georgiah): cleanup intermediate state on failure
> ```

@georgiah I'd love to see this done before your leave.

**Comment by ajwerner** [comment]
I'm sad I didn't catch the TMPDIR thing the first time around. Thanks again for investigating this and whipping up a fix so quickly.

**Review by dt — COMMENTED** [review]
👍 on commits 2 and 3, but first commit is explicitly _not_ the intended behavior.

File: pkg/blobs/local_storage.go [hunk 1]
```diff
@@ -39,9 +40,22 @@ func newLocalStorage(externalIODir string) (*localStorage, error) {
 
 func (l *localStorage) prependExternalIODir(path string) (string, error) {
 	localBase := filepath.Join(l.externalIODir, path)
-	// Make sure we didn't ../ our way back out.
+	// Whilst filepath.Join does some pruning by eliminating occurrences
+	// of "../" in the text of the path, there can still be symlinks in
+	// there, or invalid path components like the special names on
+	// windows. We really want to check upfront that the path is valid.
+	// filepath.Abs() will do this for us.
+	absPath, err := filepath.Abs(localBase)
```
[the remark this answers is not in the document]
Comment (knz) [reply pkg/blobs/local_storage.go]: I'll add the missing tests and comments then. Thanks for explaining.

**Comment by knz** [comment]
Another fun find as I am re-checking this: the current code leaks files in the temp dir if the copy fails. It forgets to remove the original file when it's done...

**Comment by ajwerner** [comment]
> Another fun find as I am re-checking this: the current code leaks files in the temp dir if the copy fails. It forgets to remove the original file when it's done...

IIRC this was known and had a TODO.

**Comment by dt** [comment]
nit: commit 2 headline could ideally mention what it fixed, eg. `blobs: fix ReadFile() handling of directories`

**Review by dt — APPROVED** [review]

**Comment by knz** [comment]
Q: is the WriteFile() method ever used to overwrite existing files?

**Comment by bdarnell** [comment]
> Another fun find as I am re-checking this: the current code leaks files in the temp dir if the copy fails. It forgets to remove the original file when it's done...

The `O_TMPFILE`/`linkat(2)` pattern could be used to avoid these leaks (create an unnamed file on the target filesystem, and don't give it a name until you're done writing to it). However, it's not portable, so we'd continue to need an alternative cleanup path. It's linux-only, and needs a relatively recent kernel (3.11, which notably excludes RHEL 7).

**Comment by knz** [comment]
I'm revisiting this code right now to put some defers in place as a 80% solution. Will ping when the PR is updated.

**Comment by knz** [comment]
Added one more commit to fix the temp file leakage, PTAL.

File: pkg/blobs/local_storage.go [hunk 2]
```diff
@@ -37,11 +37,18 @@ func newLocalStorage(externalIODir string) (*localStorage, error) {
...
+//
+// Note that we purposefully only rely on the simplified cleanup
+// performed by filepath.Join() - which is limited to stripping out
+// occurrences of "../" - because we intendedly want to allow
+// operators to "open up" their I/O directory via symlinks. Therefore,
+// a full check via filepath.Abs() would be inadequate.
 func (l *localStorage) prependExternalIODir(path string) (string, error) {
 	localBase := filepath.Join(l.externalIODir, path)
-	// Make sure we didn't ../ our way back out.
+	// Make sure no symlinks accessed a way out.
```
Comment (dt) [inline pkg/blobs/local_storage.go]: isn't this comment the opposite of what the code / comment above is doing?

**Review by dt — APPROVED** [review]

File: pkg/blobs/local_storage.go (hunk 2 above)
Comment (knz) [reply pkg/blobs/local_storage.go]: woops. Corrected during the rebase.

**Comment by knz** [comment]
TFYRs!

bors r+

**Pull request merged**